server/bin/pbench-restore-and-unpack-tarballs.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python3
+# -*- mode: python -*-
+
+"""Pbench Tar Balls & Unpack
+
+The goal is to find all tar balls that are in the backup tree and restore them
+to the archive tree, unpacking as appropriate (left to the unpack process to
+see if the tar ball is not too old), marking them as being already backed up.
+
+The general algorithm is:
+
+  1. Find all the tar balls on the backup drive
+
+  2. Sort them by the date in the tar ball from newest to oldest
+
+  3. For each tar ball
+
+     a. Check if tar ball exists in archive tree, if so, skip
+
+     b. Copy tar ball to archive tree, and .md5 and run md5sum check
+
+        * Create controller directory if it doesn't exist
+
+        * Create TO-RE-UNPACK and BACKED-UP directories if they don't exist
+
+     c. Create symlink to tar ball in TO-RE-UNPACK directory
+
+  4. After processing 10 tar balls, wait for # of TO-RE-UNPACK to drop to zero
+"""
+
+import sys
+import os
+import re
+import subprocess
+from argparse import ArgumentParser
+from datetime import datetime
+from pathlib import Path
+
+import pbench
+
+
+_NAME_ = "pbench-restore-and-unpack-tarballs"
+
+tb_pat_r = r"\S+_(\d\d\d\d)[._-](\d\d)[._-](\d\d)[T_](\d\d)[._:](\d\d)[._:](\d\d)\.tar\.xz\.md5"
+tb_pat = re.compile(tb_pat_r)
+
+
+def gen_list(backup):
+    """gen_list - traverse the given BACKUP hierarchy looking for all tar balls
+    that have a .md5 file.
+    """
+    with os.scandir(backup) as backup_scan:
+        for c_entry in backup_scan:
+            if c_entry.name.startswith(".") and c_entry.is_dir(follow_symlinks=False):
+                continue
+            if not c_entry.is_dir(follow_symlinks=False):
+                continue
+            # We have a controller directory.
+            with os.scandir(c_entry.path) as controller_scan:
+                for entry in controller_scan:
+                    if entry.name.startswith(".") and entry.is_dir(
+                        follow_symlinks=False
+                    ):
+                        continue
+                    if entry.is_dir(follow_symlinks=False):
```

Comment:
Why are we checking entry.is_dir(follow_symlinks=False): 2 times?

Reply:
Fixed.